Python/ceval.c

```diff
@@ -3691,22 +3693,31 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
             STACK_SHRINK(2);
             Py_DECREF(left);
             Py_DECREF(right);
-            assert(opcode == POP_JUMP_IF_TRUE || opcode == POP_JUMP_IF_FALSE);
+            assert(opcode == POP_JUMP_FORWARD_IF_FALSE ||
+                   opcode == POP_JUMP_BACKWARD_IF_FALSE ||
+                   opcode == POP_JUMP_FORWARD_IF_TRUE ||
+                   opcode == POP_JUMP_BACKWARD_IF_TRUE);
             int jump = (1 << (sign + 1)) & when_to_jump_mask;
             if (!jump) {
                 next_instr++;
                 NOTRACE_DISPATCH();
             }
             else {
-                JUMPTO(oparg);
+                if (opcode == POP_JUMP_BACKWARD_IF_FALSE ||
```

Comment:
There are a lot of branches here, and we are checking the eval_breaker on forward branches.

If we shift `when_to_jump_mask` left by 3 bits for backwards jumps and change 
`int jump = (1 << (sign + 1)) & when_to_jump_mask;` to
`int jump = (9 << (sign + 1)) & when_to_jump_mask;` 
then `jump` is non-zero is if we need to jump and `jump >= 8` for backward branches.

```
int jump = (9 << (sign + 1)) & when_to_jump_mask;
next_instr++;
if (jump) {
    if (jump >= 8) {
        JUMPBY(-oparg);
        CHECK_EVAL_BREAKER();
    }
    else {
        JUMPBY(oparg);
    }
}
```
Does that sound sensible to you?

Reply:
Same applies to the `int` and `str` specializations.

Reply:
Yes, nifty. I guess we could save an if:

```
if (jump >= 8){
// backward
}
else if (jump) {
// forward
}
else {
// no jump
}
```